### DISCUSSION TITLE
Immutable Dtos

### DISCUSSION THREAD

**Discussion by kamilbaczek** [kgrzybek/modular-monolith-with-ddd#142] [Ideas] [answered]
Hi, What do you think about make Dtos fetched by dapper immutable? Like this:
![image](https://user-images.githubusercontent.com/74410956/100115374-b99aa000-2e72-11eb-857f-96c89ec2b1a0.png)

**Comment by AndreiGanichev** [comment]
Hi, @kamilbaczek!
DTOs is being returned by Read Model to Presentation layer doesn't need such constraint because their only goal is to pass peace of data from database to UI. So for these classes convenience to use is more important then immutability imo. Such DTOs may be quite large because they reflect needs of UI. Sometimes they are not a flat structures but have One-To-One or One-To-Many relations and we can't map them at once. Dapper has [Multi Mapping](https://github.com/StackExchange/Dapper#multi-mapping) feature to map a sql query result to a complex DTO but we must provide public setters.

On the other hand Read Model can be used to provide data to Write Model (instead of loading aggregates). Performance is the main reason for this approach. For example [`PriceListItemData`](https://github.com/kgrzybek/modular-monolith-with-ddd/blob/master/src/Modules/Payments/Domain/PriceListItems/PriceListItemData.cs) is provided by Read Model but **declared and used in Domain layer**. But in this case `PriceListItemData` isn't used by Dapper and immutable.

**Comment by kgrzybek** [comment, marked as answer]
Hi @kamilbaczek 

We can do that but from what I remember you need to have a constructor with **parameters in a special order**. If you have different order you will have a runtime error during deserialization. Maybe something changed since I had checked this. I will check.

**Reply by kamilbaczek** [reply]
Hi @kgrzybek you are right. Parameters in a special order is required.
